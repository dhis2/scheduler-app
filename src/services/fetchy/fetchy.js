const fallbackMessage =
    'Something went wrong, but no errormessage was provided.'

const fetchy = async (url, options = {}) => {
    const settings = {
        ...options,
        credentials: 'include',
        headers: {
            'x-requested-with': 'XMLHttpRequest',
            accept: 'application/json',
            ...options.headers,
        },
    }
    let response

    // Handle fetch errors
    try {
        response = await fetch(url, settings)
    } catch (error) {
        throw new Error(
            'Something went wrong whilst retrieving data. Check if you have a working internet connection.'
        )
    }

    // Handle successful responses
    if (response.ok) {
        try {
            const data = await response.json()

            return data
        } catch (e) {
            throw new Error(
                'The response succeeded, but contained invalid JSON.'
            )
        }
    }

    // Handle error responses
    const error = new Error('')
    error.details = []

    try {
        const data = await response.json()
        const hasMessage =
            data.message &&
            typeof data.message === 'string' &&
            data.message.length > 0
        const hasDetails =
            data.response &&
            data.response.errorReports &&
            data.response.errorReports.length > 0

        if (hasMessage) {
            error.message = data.message
        }

        if (hasDetails) {
            data.response.errorReports.forEach(({ message }) => {
                if (message) {
                    error.details.push(message)
                }
            })
        }
    } catch (e) {
        throw new Error(
            'The request did not succeed and did not contain valid JSON.'
        )
    }

    if (!error.message) {
        const { statusText } = response
        const hasStatusText =
            statusText &&
            typeof statusText === 'string' &&
            statusText.length > 0

        error.message = hasStatusText ? statusText : fallbackMessage
    }

    throw error
}

export default fetchy
