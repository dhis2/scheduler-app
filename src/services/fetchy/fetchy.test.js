import 'isomorphic-fetch'
import fetchy from './fetchy'

window.fetch = jest.fn()
window.Headers = jest.fn()

describe('fetchy', () => {
    it('should reject with an error on fetch errors', () => {
        const error = new Error('Fetch error')

        window.fetch.mockImplementationOnce(() => Promise.reject(error))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should resolve with the data if the response is ok and there is data', () => {
        const expected = { data: 'data' }
        const headers = { 'Content-Type': 'application/json' }
        const init = { status: 200, statusText: 'OK', headers }
        const body = JSON.stringify(expected)
        const response = new Response(body, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).resolves.toEqual(expected)
    })

    it('should resolve with undefined if the response is ok and there is no data', () => {
        const init = { status: 200, statusText: 'OK' }
        const response = new Response(null, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).resolves.toBeUndefined()
    })

    it('should reject with an error if the response contains invalid json', () => {
        const headers = { 'Content-Type': 'application/json' }
        const init = { status: 200, statusText: 'OK', headers }
        const response = new Response([], init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should reject with an error with the message in response data if the response is not ok', () => {
        const message = 'Errormessage'
        const body = JSON.stringify({ message })
        const init = { status: 401 }
        const response = new Response(body, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrow(message)
    })

    it('should attach details to the error if there are any and the response is not ok', () => {
        const message = 'Errormessage'
        const details = 'Details'
        const data = {
            message,
            response: {
                errorReports: [
                    {
                        message: details,
                    },
                ],
            },
        }
        const body = JSON.stringify(data)
        const init = { status: 401 }
        const response = new Response(body, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return fetchy('endpoint').catch(error => {
            expect(error.details).toEqual([details])
        })
    })

    it('should reject with an error if the response is not ok and contains malformed json', () => {
        const init = { status: 401 }
        const response = new Response([], init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrowErrorMatchingSnapshot()
    })

    it('should reject with an error with the statusText if the response is not ok and there is no message', () => {
        const statusText = 'statusText'
        const body = JSON.stringify({})
        const init = { status: 401, statusText }
        const response = new Response(body, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrow(statusText)
    })

    it('should reject with an error with the fallback message if the response is not ok and there is no message or statusText', () => {
        const body = JSON.stringify({})
        const init = { status: 401, statusText: '' }
        const response = new Response(body, init)

        window.fetch.mockImplementationOnce(() => Promise.resolve(response))

        expect.assertions(1)

        return expect(fetchy('endpoint')).rejects.toThrowErrorMatchingSnapshot()
    })
})
