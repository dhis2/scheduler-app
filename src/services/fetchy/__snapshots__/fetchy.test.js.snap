// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`fetchy should reject with an error if the response is not ok and contains malformed json 1`] = `"The request did not succeed and did not contain valid JSON."`;

exports[`fetchy should reject with an error if the response is ok but contains invalid json 1`] = `"The response succeeded, but contained invalid JSON."`;

exports[`fetchy should reject with an error on fetch errors 1`] = `"Something went wrong whilst retrieving data. Check if you have a working internet connection."`;

exports[`fetchy should reject with an error with the fallback message if the response is not ok and there is no message or statusText 1`] = `"Something went wrong, but no errormessage was provided."`;
